File: 프로그래머스/lv1/172928. 공원 산책/공원 산책.java
class Solution {
    static int x, y, W, H;
    static boolean[][] map;
    static void order(char dir, int step){
        if(dir=='E' && x+step < W){
            for(int i=x+1; i<=x+step; i++){
                if(map[y][i]) return;
            }
            x += step;
        }
        else if(dir=='W' && x-step >= 0){
            for(int i=x-1; i>=x-step; i--){
                if(map[y][i]) return;
            }
            x -= step;
        }
        else if(dir=='S' && y+step < H){
            for(int i=y+1; i<=y+step; i++){
                if(map[i][x]) return;
            }
            y += step;
        }
        else if(dir=='N' && y-step >= 0){
            for(int i=y-1; i>=y-step; i--){
                if(map[i][x]) return;
            }
            y -= step;
        }
    }
    
    public int[] solution(String[] park, String[] routes) {
        int[] answer = {};
        W = park[0].length();
        H = park.length;
        map = new boolean[H][W];
        
        for(int i=0; i<H; i++){
            char[] pos = park[i].toCharArray();
            for(int j=0; j<W; j++){
                if(pos[j] == 'X'){
                    map[i][j] = true;
                }else if(pos[j] == 'S'){
                    y = i;
                    x = j;
                }
            }
        }
        
        for(String route : routes){
            char op = route.charAt(0);
            int n = route.charAt(2)-'0';
            
            order(op, n);
        }
        
        return answer = new int[] {y, x};
    }
}

File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.js
function solution(park, routes) {
    var answer = [];
    const len = park.length;
    var me;
    var map = [];
    park.forEach((str, row) => {
        var tmp = [];
        for(let col=0; col<str.length; col++){
            if(str[col] === "S") me = [row, col];
            if(str[col] === "X") tmp.push(false);
            else tmp.push(true);
        }
        map.push(tmp);
    });
    
    const op_key = {
        "N" : [0, -1, map.length],
        "S" : [0, 1, map.length],
        "W" : [1, -1, map[0].length],
        "E" : [1, 1, map[0].length],
    };
    
    function order(k, v, l, n) {
        let tmp = [...me];
        for(let i=1; i<=Number(n); i++){
            if((tmp[k] + v < 0) || tmp[k] + v >= l) {
                return;
            }
            tmp[k] += v;
            if(!map[tmp[0]][tmp[1]]){
                return;
            }
        }
        me = tmp;
        console.log("order clear "+me);
    }

    console.log(map);
    routes.forEach(route => {
        console.log(route);
        let [op, n] = route.split(" ");
        let [k, v, l] = op_key[op];
        order(k, v, l, n);
        console.log("me" + me);
    });
    
    
    
    return me;
}